radix-engine/src/blueprints/account/blueprint.rs

```diff
@@ -513,9 +561,273 @@ impl AccountBlueprint {
+        )?;
+
+        // Get the vault stored in the KeyValueStore entry - if it doesn't exist, then create it if
+        // instructed to.
+        let vault = {
+            let entry: AccountVaultIndexEntry =
+                api.key_value_entry_get_typed(kv_store_entry_lock_handle)?;
+
+            match entry {
+                Option::Some(own) => Ok(Vault(own)),
+                Option::None => {
+                    if create {
+                        let vault = Vault::create(resource_address, api)?;
+                        let encoded_value = IndexedScryptoValue::from_typed(&vault.0);
+
+                        api.key_value_entry_set_typed(
+                            kv_store_entry_lock_handle,
+                            &encoded_value.to_scrypto_value(),
+                        )?;
+                        Ok(vault)
+                    } else {
+                        Err(AccountError::VaultDoesNotExist { resource_address })
+                    }
+                }
+            }
+        };
+
+        if let Ok(mut vault) = vault {
+            match vault_fn(&mut vault, api) {
+                Ok(rtn) => {
+                    api.key_value_entry_release(kv_store_entry_lock_handle)?;
+                    Ok(rtn)
+                }
+                Err(error) => Err(error),
+            }
+        } else {
+            api.key_value_entry_release(kv_store_entry_lock_handle)?;
+            Err(vault.unwrap_err().into())
+        }
+    }
+
+    fn is_deposit_allowed<Y>(
+        deposits_mode: &AccountDepositsMode,
+        resource_address: &ResourceAddress,
+        api: &mut Y,
+    ) -> Result<bool, RuntimeError>
+    where
+        Y: ClientApi<RuntimeError>,
+    {
+        // Case: XRD - Deposit of XRD is always allowed.
+        if *resource_address == RADIX_TOKEN {
+            Ok(true)
+        } else {
+            match deposits_mode {
+                AccountDepositsMode::AllowAll => Ok(true),
+                AccountDepositsMode::AllowList(ref allow_list)
+                    if allow_list.contains(resource_address) =>
+                {
+                    Ok(true)
+                }
+                AccountDepositsMode::DisallowList(ref disallow_list)
+                    if !disallow_list.contains(resource_address) =>
+                {
+                    Ok(true)
+                }
+                // Case: Only if the resource exists (not just that we have a vault for it). So,
+                // we need to check how much of it we have. If it's more than zero then we allow
+                // it.
+                AccountDepositsMode::AllowExisting => {
+                    let amount = Self::get_vault(
+                        *resource_address,
+                        |vault, api| vault.amount(api),
+                        false,
+                        api,
+                    );
+                    if let Ok(amount) = amount {
+                        Ok(amount > Decimal::zero())
```

Comment:
My intuition here is as follows:
* My account is currently in allow only mode.
* I get a weird scammy looking resource (vault for it has been created in the account at this point)
* I get rid of this resource (perhaps by sending it to some kind of zero address or an account that I control)
* I transition my account to only accept deposits of resources that I already own. 
* If I get sent more of this scammy looking token the deposit does not go through since we're checking the balance and not the vault existence.

>  If a user has previously withdrawed all but still in AllowExisting mode, new deposits are not allowed?

They would be able to deposit if they have the required badge available and the access rule assertion passes.

Reply:
The usage scenario makes sense, especially considering that we don't have the abilitity to drop a vault.